src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionResolutionEndpoint.cs

```diff
@@ -44,23 +48,52 @@ internal class CodeActionResolutionEndpoint : ICodeActionResolutionHandler
             _resolvers = resolverMap;
         }
 
+        // Register VS LSP code action resolution server capability
+        public RegistrationExtensionResult GetRegistration() => new RegistrationExtensionResult(AssociatedServerCapability, true);
+
+        // Handle the Razor VSCode `razor/resolveCodeAction` command
         public async Task<RazorCodeActionResolutionResponse> Handle(RazorCodeActionResolutionParams request, CancellationToken cancellationToken)
         {
             if (request is null)
             {
                 throw new ArgumentNullException(nameof(request));
             }
 
-            _logger.LogDebug($"Resolving action {request.Action} with data {request.Data}.");
+            var edit = await GetWorkspaceEditAsync(request, cancellationToken).ConfigureAwait(false);
+            return new RazorCodeActionResolutionResponse() { Edit = edit };
+        }
+
+        // Handle the VS LSP `textDocument/codeActionResolve` endpoint
+        public async Task<RazorCodeAction> Handle(RazorCodeAction request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (!(request.Data is JObject paramsObj))
+            {
+                Debug.Fail($"Invalid CodeAction Received {request.Title}.");
+                return request;
+            }
 
-            if (!_resolvers.TryGetValue(request.Action, out var resolver))
+            var resolutionParams = paramsObj.ToObject<RazorCodeActionResolutionParams>();
+            request.Edit = await GetWorkspaceEditAsync(resolutionParams, cancellationToken).ConfigureAwait(false);
+            return request;
+        }
+
+        // Internal for testing
+        internal async Task<WorkspaceEdit> GetWorkspaceEditAsync(RazorCodeActionResolutionParams resolutionParams, CancellationToken cancellationToken)
+        {
+            _logger.LogDebug($"Resolving workspace edit for action `{resolutionParams.Action}` with data `{resolutionParams.Data}`.");
```

Comment:
I'd do LogInformation. This is important info. Also is "Data" readable/relevant?